src/main/java/athleticli/data/Goal.java

```diff
@@ -0,0 +1,98 @@
+import java.time.temporal.TemporalAdjusters;
+
+/**
+ * Defines the basic fields and methods for a goal.
+ */
+public abstract class Goal {
+    /**
+     * Defines different types of goal periods.
+     */
+    public enum Period {
+        WEEKLY,
+        MONTHLY
+    }
+
+    private LocalDate startDate;
+    private LocalDate endDate;
+    private Period period;
+
+    public Goal(LocalDate date, Period period) {
+        switch (period) {
+        case WEEKLY:
+            this.startDate = getFirstDayOfWeek(date);
+            this.endDate = getLastDayOfWeek(date);
+            break;
+        case MONTHLY:
+            this.startDate = getFirstDayOfMonth(date);
+            this.endDate = getLastDayOfMonth(date);
+            break;
+        default:
+        }
+        this.period = period;
+    }
+
+    /**
+     * Checks whether the date is between the period.
+     *
+     * @param date  The date to be matched.
+     * @return      Whether the date is between the period.
+     */
+    public boolean checkDate(LocalDate date) {
+        return !(date.isBefore(startDate) || date.isAfter(endDate));
+    }
+
+    /**
+     * Calculates the first day of week in which the specified date falls.
+     *
+     * @param date  The specified date.
+     * @return      The first day of week in which the specified date falls.
+     */
+    private static LocalDate getFirstDayOfWeek(LocalDate date) {
+        // manually specify Monday as the start of the week
+        // to avoid differences due to locale settings
+        return date.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY));
+    }
+
+    /**
+     * Calculates the last day of week in which the specified date falls.
+     *
+     * @param date  The specified date.
+     * @return      The last day of week in which the specified date falls.
+     */
+    private static LocalDate getLastDayOfWeek(LocalDate date) {
+        // manually specify Sunday as the end of the week
+        // to avoid differences due to locale settings
+        return date.with(TemporalAdjusters.nextOrSame(DayOfWeek.SUNDAY));
+    }
+
+    /**
+     * Calculates the first day of month in which the specified date falls.
+     *
+     * @param date  The specified date.
+     * @return      The first day of month in which the specified date falls.
+     */
+    private static LocalDate getFirstDayOfMonth(LocalDate date) {
+        return date.with(TemporalAdjusters.firstDayOfMonth());
+    }
+
+    /**
+     * Calculates the last day of month in which the specified date falls.
+     *
+     * @param date  The specified date.
+     * @return      The last day of month in which the specified date falls.
+     */
+    private static LocalDate getLastDayOfMonth(LocalDate date) {
+        return date.with(TemporalAdjusters.lastDayOfMonth());
+    }
+
+    /**
+     * Returns whether the goal is achieved.
+     *
+     * @return Whether the goal is achieved.
+     */
+    public abstract boolean isAchieved();
```

Comment:
Should this have **Data data** as a parameter or where do we best examine the data against the target value?

Reply:
> Should this have **Data data** as a parameter or where do we best examine the data against the target value?

Yes. I forgot to consider how they can access `data`.

The disadvantage of passing `data` as a parameter is that this method will have access to all data, but I failed to come up with a better way (passing in the corresponding `ArrayList` is not a good solution, because it will force us to cast the elements in it to the corresponding class).

So we might follow your suggestion and stick to what we did with the `execute` methods and pass in the entire `data` as a parameter, which is also easy to implement.

Reply:
I will fix this as soon as possible.